src/server/graphql/typesDefs/Filesystem.js

```diff
@@ -0,0 +1,113 @@
+import { gql } from "apollo-server-express";
+
+/**
+ * @typedef {Object} File
+ * @property {string} id - The UUID assigned to the file.
+ * @property {string} label - The label for the file. Usually the file name without the path.
+ * @property {string} path - Full path to the file in the data store.
+ * @property {string} permission - The maximum permission level on the file for the user. One of "read", "write", or "own".
+ * @property {number} dateCreated - File creation date as milliseconds since the epoch.
+ * @property {number} dateModified - Last modified date as milliseconds since the epoch.
+ * @property {string} type - Either "file" or "folder". Should be "file".
+ * @property {number} shareCount - The number of users the file has been shared with.
+ * @property {string} infoType - The info type of the file.
+ * @property {string} md5 - The MD5 hash of the file.
+ * @property {fileSize} number - The size of the file in bytes.
+ * @
+ */
+
+/**
+ * @typedef {Object} Folder
+ * @property {string} id - The UUID assigned to the folder.
+ * @property {string} label - The label for the folder. Usually the folder name without the path.
+ * @property {string} path - Full path to the folder in the data store.
+ * @property {string} permission - The maximum permission level on the folder for the user. One of "read", "write", or "own".
+ * @property {number} dateCreated - Folder creation date as milliseconds since the epoch.
+ * @property {number} dateModified - Last modified date as milliseconds since the epoch.
+ * @property {string} type - Either "file" or "folder". Should be "folder".
+ * @property {number} shareCount - The number of users the folder has been shared with.
+ * @property {number} fileCount - The number of files contained in the folder.
+ * @property {Array<Object>} listing - A list of files and folders contained in the folder.
```

Comment:
Oh wait, I think IntelliJ is tricking me, and both the `dir-count` and `file-count` fields for folders are used in the details panel in the current UI.

Reply:
I think that some of these are going to have to get added to the return value from the stat endpoint. Namely, `badName`, `hasSubDirs`, and `isFavorite`. They're currently only presented through the the paged-directory endpoint, which makes them hard to get at for non-listing purposes.

`files` and `folders` have been replaced by the `contents` field of the Folder type.

Unless I'm missing something, I think `total` is redundant since both `fileCount` and `folderCount` are included, so total can be computed. That's better done in the UI logic since computing it in Apollo means introducing a dependency between two fields, which is frowned upon and painful to implement in resolvers (AFAIK). 

Not sure how to handle `totalBad` (which sounds like early 90's slang to me).

Reply:
> I think that some of these are going to have to get added to the return value from the stat endpoint. Namely, `badName`, `hasSubDirs`, and `isFavorite`. They're currently only presented through the the paged-directory endpoint, which makes them hard to get at for non-listing purposes.
> 

Good point.

> Unless I'm missing something, I think `total` is redundant since both `fileCount` and `folderCount` are included, so total can be computed. That's better done in the UI logic since computing it in Apollo means introducing a dependency between two fields, which is frowned upon and painful to implement in resolvers (AFAIK).
> 

The `total` field is specific to the listing endpoint and is required for paging in the UI. It can also vary if the query is doing any filtering (just return files I own), although I don't think we currently support filtering.

> Not sure how to handle `totalBad` (which sounds like early 90's slang to me).

🤣
I'm can't remember how this is getting used currently, but it's related to the `badName` flags on any of the folder's contents. I'm not sure if we'll need that in the new UI, so it can be added later if we need it later.